@trilinos/tpetra: Implementation of threaded CrsMatrix pack

tjfulle opened the pull request:
@mhoemmen

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+    return ((!out_of_bounds_error) && (!packing_error));
+  }
+
+  KOKKOS_INLINE_FUNCTION std::string summary() const
```
You don't need to call this in a thread-parallel way, and it shouldn't be (because it does I/O), so don't mark it `KOKKOS_INLINE_FUNCTION`.

tjfulle replied:
ok

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  }
+};
+
+template<class ViewType, class OffsetsViewType,
```
See above.  ViewType needs a more descriptive name.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  LocalMatrixType local_matrix;
+  LocalMapType local_col_map;
+
+  PackCrsMatrixFunctor(ViewType num_packets_per_lid_, OffsetsViewType offsets_,
```
See above.  Trailing underscores are for private instance data.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+
+  KOKKOS_INLINE_FUNCTION void init(value_type& dst) const
+  {
+    dst.bad_index = std::numeric_limits<size_t>::max();
```
`std::numeric_limits` functions are not marked as CUDA device functions.  Use `Tpetra::Details::OrdinalTraits` instead.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  KOKKOS_INLINE_FUNCTION void
+  join (volatile value_type& dst, const volatile value_type& src) const
+  {
+    if (src.bad_index < dst.bad_index) {
```
The right join predicate is "if either dst or src shows an error."

tjfulle replied:
if `src.bad_index < dst.bad_index`, then `src` necessarily shows an error (since `bad_index` is initialized to the same value in each) and its error code and data should be transferred to `dst` so that `dst` reflects the error in the first bad index.  If `dst.bad_index <= src.bad_index` (not sure how they could ever be equal, but I include it anyway), than the error code and data in `dst` can remain unchanged.  Did I miss a case?  Or, does `dst` always have to be explicitly set?

tjfulle replied:
I concede that Kokkos may be expecting some treatment of `src` and `dst` that I don't know about.  My reasoning above is that `dst` should always contain error codes and data associated with the smallest `bad_index`.

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(LO i, value_type& dst) const
```
The parallel loop index `i` should be `const` here.

tjfulle replied:
ok

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+    const LO export_lid = export_lids[i];
+    const size_t buf_size = exports.size();
+
+    // FIXME (mfh 24 Mar 2017) Everything here assumes UVM.  If
```
Remove the FIXME comment.  This is a functor now; the caller controls where it runs.

tjfulle replied:
ok

mhoemmen commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+        dst.packing_error = ! packCrsMatrixRow(local_matrix, local_col_map,
+            num_ent_beg, val_beg, ind_beg, num_ent, export_lid);
+      }
+      if (dst.out_of_bounds_error || dst.packing_error) {
```
This should do to `dst` exactly what `join` does.

tjfulle replied:
See comments above.  I wrote the `join` and `operator` treatment of the `dst` object such that `dst` always reflected the error code and data associated with the smallest `bad_index`.

tjfulle commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  ExportLIDsViewType export_lids;
+  row_map_type row_map;
+
+  NumPacketsFunctor(ViewType num_packets_, ExportLIDsViewType export_lids_,
```
Right.  It seemed a convention (like in the [Kokkos sparse CrsMatrix](https://github.com/kokkos/kokkos-kernels/blob/master/src/Kokkos_Sparse_CrsMatrix.hpp#L124)) had trailing underscores on input arguments that had the same name as (public) member data, to avoid shadowing names.  I hadn't explicitly made the member data of the struct private, so I adopted that convention.  I can swap it if it fits better with Tpetra

tjfulle commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+  NumPacketsFunctor(ViewType num_packets_, ExportLIDsViewType export_lids_,
+      LocalMatrixType local_matrix):
+    num_packets(num_packets_), export_lids(export_lids_),
+    row_map(local_matrix.graph.row_map) {};
```
doh!

tjfulle commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+
+  KOKKOS_INLINE_FUNCTION
+    void operator() (size_t i) const {
+      LO export_lid = static_cast<LO>(export_lids[i]);
```
right :)

tjfulle commented on packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp:
```diff
@@ -70,17 +72,181 @@ class Distributor;
+///   used to store the LIDs of local rows; the input array of this function.
+/// \tparam LocalMatrixType Type of the KokkosSparse::CrsMatrix local matrix
+///
+template<class GO, class ViewType, class ExportLIDsViewType, class LocalMatrixType>
```
Right, it is _much_ cleaner this way (your suggested way, that is)!

tjfulle commented:
@mhoemmen, I addressed most of your comments, and added some of my own.  the remote-pull-test-push adds extra merge commits.  After we iterate and the PR is ready to merge, I'll first squash all the commits in to one.

mhoemmen commented:
@tjfulle I'm testing this now with Intel compiler + OpenMP.

tjfulle commented:
If you haven't started yet, I can squish all the commits in to one.

mhoemmen commented:
running tests now

mhoemmen commented:
All the tests pass.  Could you squash?  I'd like to work on it a bit :)

mhoemmen commented:
I'm uploading a partial patch that fixes some error checking issues.  Not sure if it checks what it's supposed to, but at least error checking is more sane.
[0001-TMP-Tpetra-packCrsMatrix-test-partial-patch.txt](https://github.com/trilinos/Trilinos/files/1006607/0001-TMP-Tpetra-packCrsMatrix-test-partial-patch.txt)

mhoemmen commented:
I'm redoing CUDA tests now.

tjfulle commented:
@mhoemmen, I don't remember closing this, did you?  It also shows as merged???  I did do some heavy rebasing/squashing/rewriting history, so maybe that fooled github?  At any rate, the remote checkin test script is running for intel MPI_DEBUG with and without complex as we speak so a new commit should show up in a bit.  I'll reopen then.

tjfulle commented:
The latest commits fix a bug in determining the number of entries in a row.  Previously, the number of entries for a lol row was found without checking whether the local row # > the total number of rows.  (I believe the original code did check, but after rearranging the order of some ops for the threaded version of pack, that check came later than it needed to be).  The new test that sends bad inputs found the error :)

Build failed with intel (built with gcc), looking in to it now.

mhoemmen commented:
@tjfulle Not sure what happened -- I didn't close it.  Does GitHub automatically close PRs when the commit history changes?

Glad we helped find a bug :-)

tjfulle commented:
@mhommen, apparently I rewrote enough of the history that github didn't recognize the history that was leftover.

I just pushed a new commit with the changes.  All tests pass for the standard CI tests.  I'm running the complex tests (without `FLOAT`) right now.  I committed a new script `sampleScripts/checkin-test-sems-intel.sh`.  Perhaps you could modify it to define a "standard" set of @trilinos/tpetra tests to be run before opening a PR?

tjfulle commented:
Arghh, looks like github is listing all of the commits I rebased on top of as part of the PR...

tjfulle commented:
@mhoemmen , I can close this PR and open a new one that includes only the last commit, if that is helpful.

tjfulle commented:
Okay, intel+complex test is done:

```
Enabled Packages: TpetraCore
Disabled Packages: PyTrilinos,Claps,TriKota,FEI,Moertel,STK,SEACAS,ThreadPool,OptiPack,Rythmos,Intrepid,ROL,Panzer

Build test results:
-------------------
0) MPI_RELEASE_DEBUG_SHARED_PT => Test case MPI_RELEASE_DEBUG_SHARED_PT was not run! => Does not affect push readiness! (-1.00 min)
1) MPI_RELEASE_DEBUG_COMPLEX_INTEL => passed: passed=126,notpassed=0 (17.51 min)

*** Commits for repo :

1) MPI_RELEASE_DEBUG_COMPLEX_INTEL Results:
-------------------------------------------

 passed: Trilinos/MPI_RELEASE_DEBUG_COMPLEX_INTEL: passed=126,notpassed=0

 Thu May 18 17:43:27 MDT 2017

 Enabled Packages: TpetraCore
 Disabled Packages: PyTrilinos,Claps,TriKota,FEI,Moertel,STK,SEACAS,ThreadPool,OptiPack,Rythmos,Intrepid,ROL,Panzer
 Hostname: ceerws1106
 Source Dir: /scratch/tjfulle/Developer/CI/Trilinos/cmake/tribits/ci_support/../../..
 Build Dir: /scratch/tjfulle/Developer/CI/Trilinos/CHECKIN/MPI_RELEASE_DEBUG_COMPLEX_INTEL

 CMake Cache Varibles: -DTrilinos_TRIBITS_DIR:PATH=/scratch/tjfulle/Developer/CI/Trilinos/cmake/tribits -DTrilinos_ENABLE_TESTS:BOOL=ON -DTrilinos_TEST_CATEGORIES:STRING=BASIC -DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF -DDART_TESTING_TIMEOUT:STRING=400.0 -D Trilinos_ENABLE_EXPLICIT_INSTANTIATION:BOOL=ON -D BUILD_SHARED_LIBS:BOOL=ON -D Trilinos_SHOW_DEPRECATED_WARNINGS:BOOL=ON -D Trilinos_ENABLE_Fortran:BOOL=OFF -D TPL_ENABLE_yaml-cpp:BOOL=ON -D TPL_ENABLE_SuperLU:BOOL=ON -D TPL_ENABLE_Zlib:BOOL=ON -D TPL_ENABLE_Netcdf:BOOL=ON -D TPL_ENABLE_HDF5:BOOL=ON -D TPL_ENABLE_ParMETIS:BOOL=ON -D TPL_ENABLE_Boost:BOOL=ON -D TPL_ENABLE_BoostLib:BOOL=ON -D Trilinos_CXX11_FLAGS:STRING="-std=c++11" -D CMAKE_CXX_FLAGS:STRING="-Wall" -D Trilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/sems/SEMSDevEnv.cmake -D CMAKE_BUILD_TYPE:STRING=RELEASE -D TPL_ENABLE_MPI:BOOL=ON -D Kokkos_ENABLE_DEBUG:BOOL=ON -D Teuchos_ENABLE_DEBUG:BOOL=ON -D Teuchos_ENABLE_FLOAT:BOOL=OFF -D Teuchos_ENABLE_COMPLEX:BOOL=ON -D Tpetra_INST_FLOAT:BOOL=OFF -D Tpetra_INST_COMPLEX_FLOAT:BOOL=OFF -D Tpetra_INST_COMPLEX_DOUBLE:BOOL=ON -DTrilinos_ENABLE_TpetraCore:BOOL=ON -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF -DTrilinos_ENABLE_Claps:BOOL=OFF -DTrilinos_ENABLE_TriKota:BOOL=OFF -DTrilinos_ENABLE_FEI:BOOL=OFF -DTrilinos_ENABLE_Moertel:BOOL=OFF -DTrilinos_ENABLE_STK:BOOL=OFF -DTrilinos_ENABLE_SEACAS:BOOL=OFF -DTrilinos_ENABLE_ThreadPool:BOOL=OFF -DTrilinos_ENABLE_OptiPack:BOOL=OFF -DTrilinos_ENABLE_Rythmos:BOOL=OFF -DTrilinos_ENABLE_Intrepid:BOOL=OFF -DTrilinos_ENABLE_ROL:BOOL=OFF -DTrilinos_ENABLE_Panzer:BOOL=OFF
 Make Options: -j16 
 CTest Options: -j16 

 Pull: Not Performed
 Configure: Passed (0.60 min)
 Build: Passed (16.64 min)
 Test: Passed (0.26 min)

 100% tests passed, 0 tests failed out of 126

 Label Time Summary:
 Tpetra    =  59.17 sec (126 tests)

 Total Test time (real) =  15.52 sec

 Total time for MPI_RELEASE_DEBUG_COMPLEX_INTEL = 17.51 min
```

tjfulle commented:
@mhoemmen , I'm not sure if the changes requested reflect the most recent PR?

mhoemmen commented:
@tjfulle Do you mean the MueLu-related commits?

tjfulle commented:
@mhoemmen, no, the commits before the MueLu commits.  I addressed your previous comments and retested with intel+complex.  But, since I rewrote the history, your previously requested changes point to outdated files.

mhoemmen commented:
@tjfulle That's OK -- is it safe to pull this PR and try it out, though?

tjfulle commented:
@mhoemmen, it is safe to pull and try out.  I opened a separate PR for the MueLu stuff, so you could exclude it.

mhoemmen commented:
@tjfulle I'm almost done.  Had to make some changes to combine count and offset computation, and for CUDA correctness.

mhoemmen commented:
I just pushed a revised version of this to develop: https://github.com/trilinos/Trilinos/commit/f6353cf11175b73d88c7b9e29dddbd86165c5434

Thanks @tjfulle ! :-D
